Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -3155,12 +3163,13 @@ static inline void __CGGradientInsertTransparentColor(std::vector<D2D1_GRADIENT_
     // effect based on the extend mode).   We support that by inserting a point (with transparent color) close to the start/end points, such
     // that d2d will automatically extend the transparent color, thus we obtain the desired effect for CGGradientDrawingOptions.
 
-    if (!(options & kCGGradientDrawsBeforeStartLocation)) {
+    if (!(options & kCGGradientDrawsBeforeStartLocation) && (gradientCount > 1) && !radialGradient) {
```

Comment:
Some comments on why radial gradient is special would be nice.  It may even be easier to read to separate these in to two different functions - one for radial gradient and one for linear.